zip-0222.rst

```diff
@@ -0,0 +1,322 @@
+====================== ==================================== =====================
+``tze_id``             extension ``type``                   ``compactSize``
+``tze_mode``           extension ``mode``                   ``compactSize``
+``tze_data_len``       length of precondition/witness data  ``compactSize``
+``tze_data``           serialized precondition/witness data ``witness_len`` bytes
+====================== ==================================== =====================
+
+TZE Input Encoding:
+
+====================== ==================================== ===============
+Field                  Description                          Type           
+====================== ==================================== ===============
+``prevout_hash``       hash of previous txid                ``uint256``    
+``prevout_in``         index into previous txn's outputs    ``uint32``
+``witness``            witness for prevout's precondition   ``tze_data``
+====================== ==================================== ===============
+
+
+TZE Output Encoding:
+
+====================== ==================================== ===============
+Field                  Description                          Type           
+====================== ==================================== ===============
+``amount``             spendable amount, in zatoshi         ``int64``    
+``precondition``       precondition encumbering ``amount``  ``tze_data``
+====================== ==================================== ===============
+
+
+Consensus rules
+---------------
+
+Once this ZIP becomes active, the following new consensus rules are enforced:
+
+- For each ``(outpoint, witness)`` pair in ``tze_inputs``:
+
+  - ``outpoint`` MUST reference a precondition of the same type and mode in an already-mined
+    transaction.
+  - ``tze_verify(mode, precondition, witness, context)`` MUST return ``true``.
+
+- If a transaction has non-empty ``tze_inputs`` and non-empty ``tze_outputs``, then every
+  element in both fields MUST have the same ``type``.
+
+- Non-coinbase transactions MUST have at least one of the following:
+  - nonempty transparent inputs
+  - nonempty shielded inputs
+  - nonempty ``tze_inputs``
+
+- Transactions MUST have at least one of the following:
+  - nonempty transparent outputs
+  - nonempty shielded outputs
+  - nonempty ``tze_outputs``
+
+- All ``amount`` field values of ``tze_output`` records MUST be nonnegative and not greater
+  than ``MAX_MONEY``.
+
+- The sum of amounts associated with all transparent ouputs and TZE outputs within 
+  a transaction MUST NOT exceed the sum of the amounts associated with transaction inputs
+  plus the sapling ``valueBalance`` amount.
+
+Changes to signatures over transaction digests
+----------------------------------------------
+
+The newly added parts of the transaction will be included in transaction 
+digests for signatures. See [#zip-0244]_ for the specification of the new
+digests.
```

Comment:
To ensure non-malleability, we don't want to include non-witnesses in the transaction id (the hash).